### ISSUE TITLE
segmentation fault on parallel code

### ISSUE THREAD

**Issue by bo-tato** [MoarVM/MoarVM#1864]
I get a segmentation fault on 5-10% of runs with this script:
```
use HTTP::UserAgent;


my @requests;
for ^200 -> $i {
    @requests.push: start {
        my $ua = HTTP::UserAgent.new;
        $ua.timeout = 10;
        my $response = $ua.get("https://httpbin.org/delay/2");

        say "got successful response: $i";
        CATCH {
            default {
                say "error on response: $i";
                .Str.say;
            }
        }
    }
}

say "sent all requests";
await @requests
```
This is with the latest rakudo star (v2024.10) on arch linux x86_64. I also tested on another ubuntu computer with the older version of raku in it's repository (v2022.12) and got segmentation fault and bus error there also. See https://github.com/rakudo/rakudo/issues/5683 for previous discussion of the issue where it was suggested to report here.
Aside from the segmentation fault it also often crashes with exceptions in URI:
https://gist.github.com/bo-tato/76c4fbd755c57990838e1d0fc42eeef7
https://gist.github.com/bo-tato/33fb1bd03f2e7bbf35572ba7e4622a95
I don't know if that is the same underlying issue causing the segmentation faults or some unrelated bug in URI/HTTP::UserAgent

**Comment by patrickbkr** [comment]
I haven't looked yet, but my bets are, that it's a bug in https://raku.land/github:sergot/IO::Socket::SSL (a binding of OpenSSL using NativeCall). That library isn't as stable as it should be. An easy way to verify would be to try to reproduce the errors with a non HTTPS page.

**Comment by bo-tato** [comment]
You're right it looks like a bug in IO::Socket::SSL, I tried on a http page and there's no crashes. Should I close this and open a bug there?

**Comment by patrickbkr** [comment]
That would be great. If it's your thing, you can try to golf the case a bit more (knowing that it's the SSL part at fault).

**Comment by bo-tato** [comment]
ok, yea I reduced a little to just SSL connect:
```
use IO::Socket::SSL;

my @requests;
for ^200 -> $i {
    @requests.push: start {
        my IO::Socket $ssl = IO::Socket::SSL.new(:host<github.com>, :port(443));

        say "connected: $i";
    }
}

say "sent all requests";
await @requests
```
and it's getting still getting exceptions and memory corruption. I once had a bug with openssl bindings in rust, where it worked fine highly concurrent async tokio but single threaded, worked fine async tokio multi threaded with glibc, and got occasional memory corruption with async tokio multithreaded and musl libc, and I got absolutely nowhere trying to debug it. So I think that's as far as I'll go with this one and focus my time on other bugs I have a chance of solving, though I am quite interested what is the problem if someone manages to get to the bottom of it.
Now moved to IO::Socket::SSL issues: https://github.com/sergot/io-socket-ssl/issues/25
